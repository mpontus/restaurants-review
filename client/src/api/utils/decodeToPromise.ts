import * as t from "io-ts";
import { failure } from "io-ts/lib/PathReporter";

/**
 * Validate input against io-ts schema and return result as a promise
 */
export const decodeToPromise = <A, O, I>(
  validator: t.Type<A, O, I>,
  input: I
): Promise<A> =>
  validator
    .decode(input)
    .fold(
      errors => Promise.reject(new Error(failure(errors).join("\n"))),
      value => Promise.resolve(value)
    );
