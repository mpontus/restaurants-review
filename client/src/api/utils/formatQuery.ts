import qs from "querystring";

/**
 * Filter undefined values from object
 */
const filterObject = (obj: Record<string, any>): object =>
  Object.keys(obj).reduce(
    (acc, key) => (obj[key] === undefined ? acc : { ...acc, [key]: obj[key] }),
    {}
  );

/**
 * Format query string from object after filtering undefined values
 */
export const formatQuery = (query: object) => qs.stringify(filterObject(query));
