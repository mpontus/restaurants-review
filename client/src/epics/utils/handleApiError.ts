import { of, OperatorFunction } from "rxjs";
import { catchError } from "rxjs/operators";
import { RequestError } from "../../models/RequestError";

/**
 * Helper function to handle API error
 *
 * Wraps error caught from source observable into RequestError and
 * passes it onto specified action creator before merging it into the
 * result stream.
 */
export const handleApiError = <T, R>(
  failureActionCreator: (e: any) => R
): OperatorFunction<T, T | R> =>
  catchError(error =>
    of(failureActionCreator(RequestError.fromApiError(error)))
  );
