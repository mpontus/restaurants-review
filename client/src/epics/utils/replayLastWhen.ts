import { merge, Observable, OperatorFunction } from "rxjs";
import { withLatestFrom } from "rxjs/operators";

/**
 * Replays the last item emitted by the source observable, whenever
 * selector emits.
 */
export const replayLastWhen = <T>(
  selector$: Observable<any>
): OperatorFunction<T, T> => source$ =>
  merge(source$, selector$.pipe(withLatestFrom(source$, (_, val) => val)));
