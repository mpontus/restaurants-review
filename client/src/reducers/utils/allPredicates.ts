/**
 * Helper type
 */
type Predicate<T, T1 extends T> = (input: T) => input is T1;

/**
 * All Predicates
 *
 * Creates a predicate which matches all of the passed predicates.
 */
export function allPredicates<T, T1 extends T>(
  predicate1: Predicate<T, T1>,
  ...rest: Array<(value: T1) => boolean>
): Predicate<T, T1>;

export function allPredicates<T, T1 extends T, T2 extends T1>(
  predicate1: Predicate<T, T1>,
  predicate2: Predicate<T1, T2>,
  ...rest: Array<(value: T2) => boolean>
): Predicate<T, T2>;

export function allPredicates<T, T1 extends T, T2 extends T1, T3 extends T2>(
  predicate1: Predicate<T, T1>,
  predicate2: Predicate<T1, T2>,
  predicate3: Predicate<T2, T3>,
  ...rest: Array<(value: T3) => boolean>
): Predicate<T, T3>;

export function allPredicates(...predicates: Array<(value: any) => boolean>) {
  return (input: any) => predicates.every(predicate => predicate(input));
}
