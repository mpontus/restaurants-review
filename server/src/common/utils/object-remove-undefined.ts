/**
 * Remove undefined values from an object
 */
export const objectRemoveUndefined = <T>(
  // tslint:disable-next-line:no-any
  obj: Partial<T> & { [K in string]?: any },
): Partial<T> =>
  Object.keys(obj).reduce(
    (acc, key) =>
      obj[key] === undefined ? acc : Object.assign(acc, { [key]: obj[key] }),
    {},
  );
