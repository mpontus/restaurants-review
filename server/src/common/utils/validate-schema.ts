// tslint:disable:promise-function-async
import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

/**
 * Validate Schema Error
 *
 * Error dispatched when schema validation fails.
 */
export class ValidateSchemaError extends Error {}

/**
 * Validate object against the given schema
 *
 * Used to achieve type-safety on external objects, which shape can
 * not be trusted. Dispatches {@link ValidateSchemaError} when schema
 * validation fails.
 */
export const validateSchema = <A, O, I>(
  validator: t.Type<A, O, I>,
  input: I,
): Promise<A> =>
  validator
    .decode(input)
    .fold(
      errors =>
        Promise.reject(new ValidateSchemaError(failure(errors).join('\n'))),
      value => Promise.resolve(value),
    );
