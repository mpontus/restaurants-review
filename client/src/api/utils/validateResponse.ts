import { AxiosResponse } from "axios";
import * as t from "io-ts";
import { decodeToPromise } from "./decodeToPromise";

/**
 * Validate axios response against io-ts schema and cast it to
 * corresponding type.
 */
export const validateResponse = <A, O, I>(schema: t.Type<A, O, I>) => (
  response: AxiosResponse<I>
): Promise<A> => decodeToPromise(schema, response.data);
