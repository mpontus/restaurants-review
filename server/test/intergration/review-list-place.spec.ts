import supertest from 'supertest';
import { initApp } from '../utils/init-app';
import { resetDb } from '../utils/reset-db';

let nestApp: any;

beforeAll(async () => {
  nestApp = await initApp();
});

afterAll(() => nestApp.close());

beforeEach(() => resetDb(nestApp));

describe('List place reviews', () => {
  describe('when place exists', () => {
    const seed = require('../seed/multiple-reviews');

    beforeEach(() => seed.run(nestApp));

    it('should succeed', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places/${seed.place.id}/reviews`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    describe('when limit is provided', () => {
      it('should succeed', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/places/${seed.place.id}/reviews?take=2`)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is provided', () => {
      it('should succeed', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/places/${seed.place.id}/reviews?skip=3`)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is not a number', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/places/${seed.place.id}/reviews?take=foo`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is negative', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/places/${seed.place.id}/reviews?take=-1`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is too large', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/places/${seed.place.id}/reviews?take=9000`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is not a number', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/places/${seed.place.id}/reviews?skip=foo`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is negative', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/places/${seed.place.id}/reviews?skip=-1`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('when place does not exist', () => {
    it('should fail', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places/25bd64b1-c768-5fa1-b4f5-3bacc700f3b9/reviews`)
        .expect(404);

      expect(response.body).toMatchSnapshot();
    });
  });
});
