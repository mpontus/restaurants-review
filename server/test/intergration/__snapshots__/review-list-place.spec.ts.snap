// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List place reviews when place does not exist should fail 1`] = `
Object {
  "error": "Not Found",
  "statusCode": 404,
}
`;

exports[`List place reviews when place exists should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Notice soon as brother.",
      "dateVisitted": "2012-05-22",
      "id": "9ea7f963-47d2-5a2d-8b4d-782680d3b1bf",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 2.67,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Party environment control quality full less painting.",
      "dateVisitted": "2012-05-22",
      "id": "0a931d01-a888-5523-86d1-c89bacc0a22b",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 4.17,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Notice soon as brother.",
      "dateVisitted": "2009-01-18",
      "id": "88b4c16c-e3c5-539d-b4b5-b158f4c23667",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 2.06,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Party environment control quality full less painting.",
      "dateVisitted": "2009-01-18",
      "id": "613c0e6f-4087-5546-ae99-e2436fd68219",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 3.84,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Have heart cover analysis carry.",
      "dateVisitted": "1996-03-20",
      "id": "cbf76a7b-14e5-5ba3-9c4f-af96f24cd72d",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 3.45,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Have heart cover analysis carry.",
      "dateVisitted": "1996-03-20",
      "id": "58d1832e-b696-5c49-ac55-9a4beced3f9b",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 1.46,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Before guess college speak.",
      "dateVisitted": "1986-03-12",
      "id": "231a909a-6ce5-5c65-a053-5d2853d02b52",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 3.35,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Before guess college speak.",
      "dateVisitted": "1979-02-22",
      "id": "07c76adf-2a28-5914-9128-d10c7d83f36b",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 2.89,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Have heart cover analysis carry.",
      "dateVisitted": "1979-02-22",
      "id": "6bc794d6-e0fe-5913-9ea7-66219afce965",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 2.25,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Have heart cover analysis carry.",
      "dateVisitted": "1973-11-06",
      "id": "700650e5-d979-5bfa-92e3-523f73036f9e",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 3.6,
    },
  ],
  "total": 11,
}
`;

exports[`List place reviews when place exists when limit is negative should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "take must not be less than 0",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": -1,
      },
      "value": -1,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List place reviews when place exists when limit is not a number should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "max": "take must not be greater than 50",
        "min": "take must not be less than 0",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": null,
      },
      "value": null,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List place reviews when place exists when limit is provided should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Party environment control quality full less painting.",
      "dateVisitted": "2012-05-22",
      "id": "0a931d01-a888-5523-86d1-c89bacc0a22b",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 4.17,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Notice soon as brother.",
      "dateVisitted": "2012-05-22",
      "id": "9ea7f963-47d2-5a2d-8b4d-782680d3b1bf",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 2.67,
    },
  ],
  "total": 11,
}
`;

exports[`List place reviews when place exists when limit is too large should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "max": "take must not be greater than 50",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": 9000,
      },
      "value": 9000,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List place reviews when place exists when offset is negative should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "skip must not be less than 0",
      },
      "property": "skip",
      "target": Object {
        "skip": -1,
        "take": 10,
      },
      "value": -1,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List place reviews when place exists when offset is not a number should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "skip must not be less than 0",
      },
      "property": "skip",
      "target": Object {
        "skip": null,
        "take": 10,
      },
      "value": null,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List place reviews when place exists when offset is provided should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Notice soon as brother.",
      "dateVisitted": "2009-01-18",
      "id": "88b4c16c-e3c5-539d-b4b5-b158f4c23667",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 2.06,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Have heart cover analysis carry.",
      "dateVisitted": "1996-03-20",
      "id": "58d1832e-b696-5c49-ac55-9a4beced3f9b",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 1.46,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Have heart cover analysis carry.",
      "dateVisitted": "1996-03-20",
      "id": "cbf76a7b-14e5-5ba3-9c4f-af96f24cd72d",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 3.45,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Before guess college speak.",
      "dateVisitted": "1986-03-12",
      "id": "231a909a-6ce5-5c65-a053-5d2853d02b52",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 3.35,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Have heart cover analysis carry.",
      "dateVisitted": "1979-02-22",
      "id": "6bc794d6-e0fe-5913-9ea7-66219afce965",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 2.25,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Before guess college speak.",
      "dateVisitted": "1979-02-22",
      "id": "07c76adf-2a28-5914-9128-d10c7d83f36b",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 2.89,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Serious inside else memory if six.",
      "dateVisitted": "1973-11-06",
      "id": "e6336264-dacd-59dc-8a16-f6e06351f47b",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 4.69,
    },
    Object {
      "author": Object {
        "name": "Norma Fisher",
      },
      "comment": "Have heart cover analysis carry.",
      "dateVisitted": "1973-11-06",
      "id": "700650e5-d979-5bfa-92e3-523f73036f9e",
      "place": Object {
        "address": "261 Alexandre Forks",
        "id": "99ea5b8b-c27d-5560-a6a3-35f23fd8e83d",
        "rating": 0,
        "title": "The Exhibit",
      },
      "rating": 3.6,
    },
  ],
  "total": 11,
}
`;
